KCM responder

jhrozek opened the pull request:
So far all the credential caches are stored in memory. However, all the
internal APIs are ready for talking to sssd-secrets and store the caches
there which is what I'm working on now.

As per testing, it should be enough to set the default ccache name to
"KCM:" and make sure the sssd-kcm socket is enabled and started and the
sssd-kcm service is enabled.

It is important to test access control. Manually, I tested two different
UNIX users, tested setting the KRB5CCNAME variable to someone else's ccache
and so on, but more testing is welcome.

jhrozek commented:
I had an IRC discussion with @simo5  about the patches and he requested some changes:
 * don't special case /kcm in the routing protocol and allow routing KCM to Custodia
 * serialize write operations from the same UIDs (or all operations from the same UIDs if that's easier for start) to avoid races between operations that need multiple HTTP round-trips
 * don't use HTTP PATCH until Custodia supports it. First, open an issue against Custodia to support PATCH and then support the same semantics

jhrozek commented:
also, store the ccaches at cn=kcm not at cn=ccaches

jhrozek commented:
also remove all the KCM-specific knowledge from sssd-secrets, like the session/peristent checks.

jhrozek commented:
I pushed new patches that implement what Simo asked for:
 * the kcm_ops module now queues all requests from the same UID
 * the secrets module does no special casing of the KCM paths
 * the KCM data is stored at cn=kcm top-level container in secrets
 * HTTP patch is no longer used, but instead delete/put is used because we now serialize the operations from the same UID

mzidek-rh commented:
On VM with were libcurl-devel is not installed, the configure succeeds, but compilation fails with:
/home/user/gitrepo/sssd/src/util/tev_curl.c:32:23: fatal error: curl/curl.h: No such file or directory
 #include <curl/curl.h>

We should check for libcurl headers in configure.

mzidek-rh commented:
Actually we do check for libcurl in configure. The reproducer is:


$ sudo dnf remove libcurl-devel
$ configure | grep curl
configure: WARNING:
The libcurl development library was not found. Some features will be disabled.

$ make
<snip>
  CC       src/util/sssd_kcm-tev_curl.o
  CC       src/responder/common/sssd_kcm-negcache_files.o
  CC       src/responder/common/sssd_kcm-negcache.o
/home/user/gitrepo/sssd/src/util/tev_curl.c:32:23: fatal error: curl/curl.h: No such file or directory
 #include <curl/curl.h>



So we try to build some parts with curl even if curl is not available.

jhrozek commented:
On Tue, Mar 21, 2017 at 09:26:23AM -0700, mzidek-rh wrote:
> Actually we do check for libcurl in configure. The reproducer is:
> 
> 
> $ sudo dnf remove libcurl-devel
> $ configure | grep curl
> configure: WARNING:
> The libcurl development library was not found. Some features will be disabled.
> 
> $ make
> <snip>
>   CC       src/util/sssd_kcm-tev_curl.o
>   CC       src/responder/common/sssd_kcm-negcache_files.o
>   CC       src/responder/common/sssd_kcm-negcache.o
> /home/user/gitrepo/sssd/src/util/tev_curl.c:32:23: fatal error: curl/curl.h: No such file or directory
>  #include <curl/curl.h>
> 
> 
> 
> So we try to build some parts with curl even if curl is not available.

Yes, this was because previously curl was only needed for tests. I
changed the curl condition to error out and moved the check for libcurl
to the section that is evaluated only if kcm is selected.

mzidek-rh commented:
Nitpick: In the contrib/kcm_default_ccache it would be good to indicate (in the comments) where the snippet belongs. It is written in the commit message, but if someone just opens the file he/she may come to a wrong conclusion that it is a sssd config snippet.

lslebodn commented on src/tests/intg/test_kcm.py:
```diff
@@ -0,0 +1,439 @@
+
+
+@pytest.fixture
+def setup_for_kcm_sec(request, kdc_instance):
```
Just for the record.

Please split this fixture into two fixtures. Because sometimes sssd-secrets was not properly cleaned-up
(probably) due to `skip` in `create_sssd_kcm_fixture`.

jhrozek commented:
On Tue, Mar 21, 2017 at 10:31:10AM -0700, mzidek-rh wrote:
> Nitpick: In the contrib/kcm_default_ccache it would be good to indicate (in the comments) where the snippet belongs.

Done.

> It is written in the commit message, but if someone just opens the file
> he/she

The gender neutral pronoun is they :-P

jhrozek commented:
I made two fixes related to documentation and packaging. They are already squashed, but to ease the review, the documentation fix was:
```
diff --git a/src/man/sssd-kcm.8.xml b/src/man/sssd-kcm.8.xml
index b28b3bf..5dc9383 100644
--- a/src/man/sssd-kcm.8.xml
+++ b/src/man/sssd-kcm.8.xml
@@ -119,6 +119,24 @@ systemctl enable sssd-kcm.service
         </para>
     </refsect1>
 
+    <refsect1 id='storage'>
+        <title>THE CREDENTIAL CACHE STORAGE</title>
+        <para>
+            The credential caches are stored in the SSSD secrets service (see
+            <citerefentry>
+                <refentrytitle>sssd-secrets</refentrytitle><manvolnum>5</manvolnum>
+            </citerefentry>
+            for more details). Therefore it is important that also the sssd-secrets
+            service is enabled and its socket is started:
+            <programlisting>
+systemctl start sssd-secrets.socket
+systemctl enable sssd-secrets.socket
+systemctl enable sssd-secrets.service
+            </programlisting>
+            Your distribution should already set the dependencies between the services.
+        </para>
+    </refsect1>
+
     <refsect1 id='options'>
         <title>CONFIGURATION OPTIONS</title>
         <para>
```

And the packaging fix was about pointing to the right socket -- I didn't notice the wrong socket name because I had the configuration on the sssd side overriden in the same manner:
```
index 45d4853..ac88fca 100644
--- a/contrib/kcm_default_ccache
+++ b/contrib/kcm_default_ccache
@@ -2,8 +2,11 @@
 # directory that is included from the Kerberos configuration file (/etc/krb5.conf)
 # On Fedora/RHEL/CentOS, this is /etc/krb5.conf.d/
 #
-# To enable the KCM credential cache, uncomment the two following lines and enable the KCM socket:
+# To enable the KCM credential cache, uncomment the following lines and
+# enable the KCM socket and the service:
 #   systemctl enable sssd-kcm.socket
+#   systemctl start sssd-kcm.socket
+#   systemctl enable sssd-kcm.service
 
 #[libdefaults]
 #    default_ccache_name = KCM:
diff --git a/src/sysv/systemd/sssd-kcm.socket.in b/src/sysv/systemd/sssd-kcm.socket.in
index b115b9a..80ec1c0 100644
--- a/src/sysv/systemd/sssd-kcm.socket.in
+++ b/src/sysv/systemd/sssd-kcm.socket.in
@@ -1,9 +1,10 @@
 [Unit]
 Description=SSSD Secrets Service responder socket
-Documentation=man:sssd-kcm(5)
+Documentation=man:sssd-kcm(8)
+Requires=sssd-secrets.socket
 
 [Socket]
-ListenStream=@localstatedir@/run/kcm.socket
+ListenStream=@localstatedir@/run/.heim_org.h5l.kcm-socket
 
 [Install]
 WantedBy=sockets.target
```

jhrozek commented:
I also removed the declaration of sss_krb5_marshal_creds which was a leftover from parsing the creds (which is something I have in a different branch until we need to renew credentials..)

jhrozek commented:
CI: http://sssd-ci.duckdns.org/logs/job/65/17/summary.html

mzidek-rh commented:
Some nitpicks mostly regarding the lack of DEBUG messages.

src/responder/kcm/kcmsrv_cmd.c
164     /* First 16 bits are 8 bit major and 8bit major protocol version */
                                                                        ^^^^^ Should be minor

-----------------------------------

src/responder/kcm/kcmsrv_cmd.c - this module could use more debug messages. First because it is tevent code and second because the debug messages server as "natural" comments. For example:
311 static void kcm_send_reply(struct cli_ctx *cctx,
312                            struct kcm_op_io *op_io,
313                            struct kcm_repbuf *repbuf)
314 {
315     errno_t ret;
316 
317     ret = kcm_output_construct(op_io, repbuf);
318     if (ret != EOK) {

It would be good to print DEBUG message here with the error code.

319         kcm_reply_error(cctx, ret, repbuf);
320         return;
321     }

-----------------------------------

This function:
509 static int kcm_send_data(struct cli_ctx *cctx)
Could have a debug a message after each kcm_write_iovec operation saying what it failed to write.

-----------------------------------

369 static void kcm_cmd_request_done(struct tevent_req *req)
389 static errno_t kcm_recv_data(int fd, struct kcm_reqbuf *reqbuf)

IMO the above functions should also print DEBUG messages on errors.

-----------------------------------

The function
bool kcm_cc_access(struct kcm_ccache *cc

Should IMO include TRACE_FUNC level debug messages at least for all cases when the result is false. The root cause is lost in the above calling functions and it ends up later in some async code.

-----------------------------------

1166 static errno_t sec_list_recv(struct tevent_req *req,
1167                              TALLOC_CTX *mem_ctx,
1168                              const char ***_sec_key_list,
1169                              size_t *_sec_key_list_len)
1170 
1171 {
1172     struct sec_list_state *state = tevent_req_data(req,
1173                                                 struct sec_list_state);
1174 
1175     TEVENT_REQ_RETURN_ON_ERROR(req);
1176 
1177     if (_sec_key_list != NULL) {
1178         *_sec_key_list = talloc_steal(mem_ctx, state->sec_key_list);
1179     }
1180     if (_sec_key_list_len) {
                                           ^^^^^^ use != NULL here too

1181         *_sec_key_list_len = state->sec_key_list_len;
1182     }
1183     return EOK;
1184 }

-----------------------------------

1247 static void sec_get_done(struct tevent_req *subreq)
1248 {
1249     errno_t ret;
1250     struct tevent_req *req = tevent_req_callback_data(subreq, struct tevent_req);
1251     struct sec_get_state *state = tevent_req_data(req,
1252                                                 struct sec_get_state);
1253     struct sss_iobuf *outbuf;
1254     const char *sec_value;
1255     int http_code;
1256 
1257     ret = tcurl_http_recv(state, subreq, &http_code, &outbuf);
1258     talloc_zfree(subreq);
1259     if (ret != EOK) {
1260         DEBUG(SSSDBG_OP_FAILURE,
1261               "GET HTTP request failed [%d]: %s\n", ret, sss_strerror(ret));
1262         tevent_req_error(req, ret);
1263         return;
1264     }
1265 
1266     if (http_code != 200) {
1267         ret = http2errno(http_code);

Add DEBUG message here

1268         tevent_req_error(req, ret);
1269         return;
1270     }

mzidek-rh commented:
Other then that it LGTM and what I tried worked for me.

jhrozek commented:
Thank you for the testing and the review, I hope I addressed all comments.

simo5 commented on src/responder/secrets/providers.c:
```diff
@@ -42,10 +50,43 @@ static int sec_map_url_to_user_path(struct sec_req_ctx *secreq,
     return EOK;
 }
 
+static int kcm_map_url_to_user_path(struct sec_req_ctx *secreq,
```
Can you change the function name to kcm_map_url_to_path ? We are not mapping to any "user" path in here.

jhrozek replied:
done

simo5 commented on src/responder/secrets/local.c:
```diff
@@ -738,6 +742,11 @@ static int local_secrets_map_path(TALLOC_CTX *mem_ctx,
                                      secreq->mapped_path + (sizeof(SEC_BASEPATH) - 1));
         basedn = SECRETS_BASEDN;
+    } else if (strncmp(secreq->mapped_path,
+               SEC_KCM_BASEPATH, sizeof(SEC_KCM_BASEPATH) - 1) == 0) {
```
I'm thinking we may want to do the check on euid here, rather than the url mapping function.

simo5 replied:
Or perhaps in sec_req_routing()

jhrozek replied:
There is a list of allowed prefixes that are routed to, but the euid check is still in the mapping function. It's not needed to have it there but IMO also not harmful.

simo5 commented on src/responder/kcm/kcmsrv_op_queue.c:
```diff
@@ -0,0 +1,230 @@
+    /* Remove the current entry from the queue */
+    DLIST_REMOVE(*entry->head, entry);
+
+    if (next_entry == NULL) {
```
Not sure why you are doing this only if there is no other entry.
As far as I understand, you are using the hash as a lock which needs to be freed each time a request is done so the next one can take it ?
Or am I wrong ?
Maybe a comment that describes how the queue is supposed to work will help insure it does operate according to design.

jhrozek replied:
The hash's key is the UID of the client, the value is the queue. If there is nothing in the queue at the time the request arrives (`hash[uid] == none`) then we create a request and add it to the hash table (`hash[uid] = new_request()`) but continue immediatelly. If there is already some request (`if uid in hash`) then we append the request (`hash[uid].append(new_request())`) and wait for the previous requests to finish.

I commented the code more and also added a unit test for the module, hopefully it's clearer now.

simo5 commented on src/responder/kcm/kcmsrv_ccache_secrets.c:
```diff
@@ -0,0 +1,3063 @@
+}
+
+/*
+ * ccache marshalling to JSON
```
I think it would be better to split this file and push all marshalling functions in a separate file.
would be good for reviews and general readability as this file is quite big.

jhrozek replied:
done and since there is a new file with the marshalling, I also wrote a test for the marshalling to check jansson is used correctly and there are no leaks.

jhrozek commented on src/responder/secrets/providers.c:
```diff
@@ -42,32 +43,190 @@ static int sec_map_url_to_user_path(struct sec_req_ctx *secreq,
+/*
+ * Internally, we map the incoming path to:
+ * SEC_KCM_PFX/$session/$impersonated_uid/$relpath
+ *
```
this was fixed in the previous iteration already. now only the peer credentials are checked and the patch is forwarded to the provider

jhrozek commented:
I pushed new patches. CI passed: http://sssd-ci.duckdns.org/logs/job/65/29/summary.html

jhrozek commented:
rebased on master

jhrozek commented:
CI on the rebased patches: http://sssd-ci.duckdns.org/logs/job/65/45/summary.html

mzidek-rh commented:
All my concerns have been addressed (and If I am not mistaken also Simo's). ACK from me.

simo5 commented:
All looks good to go.

jhrozek commented:
@lslebodn I pushed a new patchset that should fix the warning in the kcm_queue test, can you verify if all warnings you saw are fixed?

jhrozek commented:
I also squashed Lukas' patch to fix some python3 issues. I'm just waiting for our CI to finish.

lslebodn commented:
retest this please

jhrozek commented:
OK, hopefully final CI run: http://sssd-ci.duckdns.org/logs/job/65/48/summary.html

jhrozek commented:
Pushed to master:
 * e89ba95737202d551db2c9524127e6c4cf308796
 * 2b5518eeaacc6245cfa77ee4a7086f16208060fc
 * 35c9dfe9ba78d3a635cd1af0fb6349ba44344623
 * cac0db2f8004ae88b9263dc3888a11a2d3d3d114
 * c9db8b8b19827c3d492b8d2769aa77a37dbc12d3
 * 60612b5fbdaaa62ebe6c7f4c27200316f08506d6
 * 73ce539aa70f43ccd5302b3ef8a02ff028558b12
 * 8bb2fcbce7c3fcfd986f1bc835fbcc43ac7cd9d1
 * 0700118d8388c38b8cb28279510b206b76a3a411
 * ba89271f594e5ed381b4dcb876a2d2787cf51902
 * 1ec4198f38d34a1f82a2db55d8c9782a434fb55f
 * 70fe6e2bb398b8669ad1aebeaf0abcbffc307475
 * bea0dc79faf609de8603cb42f190adae544bc8fb
 * 9dcdbf596e138df3eec202487549a67cd3b0091b
 * b9c563c29243291f40489bb0dcbf3946fca72d58
 * 1dbf09404e20b6e30a24afe72b6d349734aee62f
 * 5f7f45a64bdb9353f15b945db4ad2564b4b28ab2
 * 4f511a4c5f0084e22ce4c7613f1b279533c68cc5
 * c194e8d7cad0184d710d9979e9f12d5cfe176f4a
 * 24889dc5e7eb7bc992ab0fa05edfdfa1d157131a
 * 3a4a88259ba90d3dc45c1adbbfd39bd7c0204a12
